crates/memory/src/memory.rs

```diff
@@ -14,7 +14,7 @@ pub struct GuestMemory<'a> {
 
 impl<'a> GuestMemory<'a> {
     pub fn new(ptr: *mut u8, len: u32) -> GuestMemory<'a> {
-        assert_eq!(ptr as usize % 4096, 0, "GuestMemory must be page-aligned");
+        // assert_eq!(ptr as usize % 4096, 0, "GuestMemory must be page-aligned");
```

Comment:
I think I still know too little about how Wasm memory is passed in and out of the runtime to say with even some degree of certainty whether we should have this assertion here or not. Anyhow, I've had to comment it out since I'm unable to force an exact alignment of `4096` in Rust. FWIW, it is possible to require Rust to allocate *at least* `4096` but it's not guaranteed not to be more (and this is where the assertion fails).

For example, I've tried:

```rust
#[repr(align(4096))]
struct HostMemory {
    buffer: Vec<u8>,
}
```

Which sometimes is properly aligned to `4096` by Rust and sometimes isn't. Having said that, do we want to check the alignment of Wasm's linear memory in host's address space, or do we only care whether the guest pointers are properly aligned and that the guest memory is a multiple of `4096` in length?

Reply:
Sorry, that assertion was a bit sloppy and is probably more aggressive than required. However, some amount of alignment is required, because GuestMemory's ptr creation will check indices into the memory to make sure they correspond to the GuestType alignment. We need everything to follow rust's alignment rules in order for GuestTypeRef to work at all, and for GuestType::validate and GuestTypeClone to be implemented with ordinary copies rather than read_unaligned/write_unaligned.

The struct you wrote isnt making sure the contents of buffer are aligned there, because Vec will allocate its contents on the heap, where alignment is much smaller (either word or double-word, I forget). I pushed a HostMemory to your branch defined with an array, rather than a vec, that will be aligned correctly.

In lucet we create hostmemory with mmap. I think we should use mmap to allocate the test hostmemory region so that we can bracket it with guard regions thatll segfault if accessed.

Reply:
Ah, OK, that makes sense (I think). I totally forgot about the fact that (obv!) `Vec` is heap-allocated!

Way ahead of you, I've got the tab with Lucet's `mmap` open and was browsing through it already ;-) I just thought that it would be good to do by-hand basic alloc stuff first before adding more mechanisms into play.

Reply:
BTW, are you sure you've pushed the `HostMemory` into the branch? I cannot see any commit here 🤔

Reply:
OK, so not to lose momentum, I've now re-enabled the alignment checks, and followed your suggestion of wrapping a slice rather than a `Vec` inside a `struct` that's then `#[repr(align(4096))]` aligned.

Reply:
My bad, I had forgotten to actually push it. Redid the commit on top of your changes and pushed.